proposals/init.md

```diff
@@ -0,0 +1,582 @@
+the expense of binary compatibility. Digging in a bit the extra safety is not
+exactly what it seems. It will not protect against the following circumstances:
+
+1. Reflection over `public` members
+1. The use of `dynamic` 
+1. Compilers that don't recognize modreqs
+
+It should also be considered that, when we complete the IL verification rules 
+for .NET 5, `init` will be one of those rules. That means extra enforcement 
+will be gained from simply verifying compilers emitting verifiable IL.
+
+The primary languages for .NET (C#, F# and VB) will all be updated to 
+recognize these `init` accessors. Hence the only realistic scenario here is 
+when a C# 9 compiler emits `init` properties and they are seen by an older 
+toolset such as C# 8, VB 15, etc ... C# 8. That is the trade off to consider
+and weigh against binary compatibility.
+
+**Note**
+This discussion primarily applies to members only, not to fields. While `init`
+fields were rejected by LDM they are still interesting to consider for the 
+modreq vs. attribute discussion. The `init` feature for fields is a relaxation
+of the existing restriction of `readonly`. That means if we emit the fields as
+`readonly` + an attribute there is no risk of older compilers mis-using the 
+field because they would already recognize `readonly`. Hence using a modreq here
+doesn't add any extra protection.
+
+**Resolution**
+The feature will use a modreq to encode the property `init` setter. The
+compelling factors were (in no particular order):
+
+* Desire to discourage older compilers from violating `init` semantics
+* Desire to make adding or removing `init` in a `virtual` declaratio or 
+`interface` both a source and binary breaking change.
+
+Given there was also no significant support for removing `init` to be a 
+binary compatible change it made the choice of using modreq straight forward.
+
+### init vs. initonly
+There were three syntax forms which got significant consideration during our
+LDM meeting:
+
+```cs
+// 1. Use init 
+int Option1 { get; init; }
+// 2. Use init set
+int Option2 { get; init set; }
+// 3. Use initonly
+int Option3 { get; initonly; }
+```
+
+**Resolution**
+There was no syntax which was overwelmingly favored in LDM.
+
+One point which got significant attention was how the choice of syntax would
+impact our ability to do `init` members as a general feature in the future.
+Choosing option 1 would mean that it would be difficult to define a property
+which had a `init` style `get` method in the future. Eventually it was decided
+that if we decided to go forward with general `init` members in future we could
+allow `init` to be a modifier in the property accessor list as well as a short
+hand for `init set`. Essentially the following two declarations would be
+identical.
+
+```cs
+int Property1 { get; init; }
+int Property1 { get; init set; }
+```
+
+The decision was made to move forward with `init` as a standalone accessor in
```

Comment:
If init is a standalone accessor, is there any desire to allow both init and set, but with different accessibility. For example you may want to allow setting a property at initialization via an object initializer instead of via a constructor, but allow setting it inside of the class.

In other words this pattern:

```csharp
class C {
    public C(int prop) => Prop = prop;
    public int Prop { get; private set; }


}
```

Could now be replaced with:

```csharp
class C {
    public int Prop { get; init; private set; }
}
```

Reply:
If I understand it correctly, as per section “Generate three accessors”, having `init` and `set` be separate was considered but the cost is apparently too much. So `init` is direct replacement for `set` if used.

Reply:
@poke
That was discussing this as an emit strategy. I'm asking whether there's a benefit to having this as a language feature. None of the benefits or drawbacks to that suggested emit strategy apply to this this suggestion.

Reply:
We discussed whether `init` accessor is a separate thing (and should use a separate slot) than `set` accessor, but landed on just re-using the same slot and not allowing having both an `init` and `set`.

https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-04-13.md#init-only-members